### PR TITLE
feat: force login (close #1132)

### LINKED ISSUE #1132: Feature: Force login when checking the box
### Type of feature

🍕 Feature

### Current behavior

Screen Shot 2023-05-01 at 9 06 47 AM

Today when a user is logged out, we do not allow them to check a box. Instead we make this an allowed action. When the user clicks "Create Insight," that should logged them in and present them with an insight page with the items selected in the cart.  

### Suggested solution

Screen Shot 2023-05-01 at 9 08 44 AM

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by a0m0rajab**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description

This PR closes #1132

## Mobile & Desktop Screenshots/Recordings

image
When the user is not logged it, it will force login.

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

**Comment by brandonroberts** [comment]
We can land this but if they select repos and are go to login, when they return the repos won't be selected

File: components/organisms/Repositories/repositories.tsx [hunk 1]
```diff
@@ -55,7 +55,15 @@ const Repositories = ({ repositories }: RepositoriesProps): JSX.Element => {
         "/hub/insights/new"
       );
     } else {
-      signIn({ provider: "github" });
+      signIn({ 
+        provider: "github",
+        options: {
+          redirectTo: `${window.location.origin}/hub/insights/new`,
+          queryParams:{
```
Comment (a0m0rajab) [inline components/organisms/Repositories/repositories.tsx:62]: This is based on the supabase definition: 

```js
export type SignInWithOAuthCredentials = {
  /** One of the providers supported by GoTrue. */
  provider: Provider
  options?: {
    /** A URL to send the user to after they are confirmed. */
    redirectTo?: string
    /** A space-separated list of scopes granted to the OAuth application. */
    scopes?: string
    /** An object of query params */
    queryParams?: { [key: string]: string }
    /** If set to true does not immediately redirect the current browser context to visit the OAuth authorization page for the provider. */
    skipBrowserRedirect?: boolean
  }
}
```
Yet the query params did not work when the page redirected to insights/new 
How can I debug it? I tried to check the network tab from dev tools in Chrome but could not find this redirect. 

Is there another tool I need to try/test?

Comment (brandonroberts) [reply components/organisms/Repositories/repositories.tsx:62]: It looks like if you pass the query params as part of the `redirectTo`, you are redirected back to the app after logging in with those query params kept intact.

```
${window.location.origin}/hub/insights/new?param=value
```

We can't pass `selectedRepos` as-is because its too much data for the URL, so they'll be some additional refactoring there

**Comment by brandonroberts** [comment]
@a0m0rajab does this redirect to the new insight page with the selected repos after login?

**Comment by a0m0rajab** [comment]
No, I just misunderstood your first comment, wanted to know what you meant by refactoring. 
I think it means something to serialize selectedRepos and send it within the URL, I will work on that.

**Comment by brandonroberts** [comment]
@a0m0rajab yes, only passing selectRepos as just the IDs would be a better fix. Then you'd have to fetch the repos using the `selectedRepos` in the `InsightPage.tsx` component and add them to the listed repos.

**Comment by a0m0rajab** [comment]
Thank you for that, Just another solution I got in mind is using localStorage to get the repos as well. 
If the query params did not work, I might work on this one.

**Comment by brandonroberts** [comment]
We discussed using local storage previously and its too prone to error

**Comment by OgDev-01** [comment]
Hey @a0m0rajab, have you been able to try @brandonroberts suggestion?... Let me know if you are blocked anytime 👍

**Comment by a0m0rajab** [comment]
@OgDev-01 Thank you for asking, I just did not find time recently, might do it at the weekend. 
@brandonroberts Thank you for letting me know, I will go with passing IDs and see how things will work.

**Comment by a0m0rajab** [comment]
Got blocked while working on this, had the same long URI issue from supabase this time.

**Comment by takanome-dev** [comment]
Hey @a0m0rajab, the Supabase redirect issue was fixed in https://github.com/open-sauced/insights/pull/1240. Try pulling the latest changes and try it again. Let us know if you're still facing an issue :pizza:

**Comment by a0m0rajab** [comment]
@takanome-dev it did not work, still the same issue, 

image


I think it was my mistake, just solved the issue, had to remove a test code.

**Comment by a0m0rajab** [comment]
This PR worked when tested on the: https://github.com/open-sauced/insights/pull/1184/commits/fef7f6042cb9c9b3a8ff5b126745ceac2af540de commit. 

After merge, it never redirected to the new page after authentication, it always redirected to the home page, it needs some investigation. 

I tried to remove the code in the middleware function to see if it's the issue, it worked but while waiting, it redirected again to the home page. I think there is a conflict with the authentication check logic and force auth.

**Comment by a0m0rajab** [comment]
Not sure why but the last issue I am facing is related to the next codes, removing them solved the issue, but I need to see what caused it. 

https://github.com/open-sauced/insights/blob/1e16bde2139c638443a53a1ac89444f93d921a61/layouts/private-wrapper.tsx#L16

https://github.com/open-sauced/insights/blob/1e16bde2139c638443a53a1ac89444f93d921a61/pages/pages/%5BuserOrg%5D/%5BfilterName%5D/%5BtoolName%5D.tsx#L27

https://github.com/open-sauced/insights/blob/1e16bde2139c638443a53a1ac89444f93d921a61/middleware.ts#L1-L29

This is related to middleware and PrivateWrapper

@takanome-dev, you worked on those pages, do you have an idea how to solve this? 

I tried to make the signin function async and add await but it did not work.

**Comment by a0m0rajab** [comment]
I added a condition to disable the middleware logic when this feature is enabled, I would love to get your thoughts and inputs. @takanome-dev

**Comment by takanome-dev** [comment]
> I added a condition to disable the middleware logic when this feature is enabled, I would love to get your thoughts and inputs

Yeah, there is some kind of conflict with the middleware. But adding the following check in the `private-wrapper` component will allow a user to stay in the `add new insight page` after logout
```tsx
if (router.asPath.includes("selectedReposIDs")) return;
```
I didn't manage to find a solution to this yet, got some thoughts on this @brandonroberts?

Btw, @a0m0rajab the `TableTitle` component does not have a `text` prop. That's why the build is failing...

**Comment by a0m0rajab** [comment]
I fixed the build, will try to investigate the logout and the signin functions.

**Comment by a0m0rajab** [comment]
I think this is happening due to not having the login information in cookies while redirecting: 
image

The private wrapper ran 3 times, till it was able to retrieve the user information. 

This is the 3rd time screenshot: 
image


And the private wrapper was called 4 times in total. (while on dev)

**Comment by a0m0rajab** [comment]
I did a quick search and found this two links: 
This [discussion on supabase repo](https://github.com/orgs/supabase/discussions/2842)is facing the same issue as what we are facing, I need read and see the solutions in it. 

This [stack over flow](https://stackoverflow.com/questions/75791108/protecting-routes-in-nextjs-using-supabase) feels a bit direct and we can use it:  

Google search query: react router supabase private link

**Comment by a0m0rajab** [comment]
I am trying the isLoading state scenario to see if this might solve the issue. 

This [article ](https://blog.logrocket.com/use-redux-next-js/)helped me to realize that state is not persistent and need to do further engineering for that, this discussing reminded me of [store idea](https://github.com/vercel/next.js/discussions/38871), this is the [library ](https://github.com/pmndrs/zustand)that we use here for global state management

**Comment by brandonroberts** [comment]
@a0m0rajab So even with only using the repository IDs, the URL is too long? 

We definitely don't want to persist the entire global state there for this use case. Storing only the `selectRepoIds` in local storage to use on return may be an acceptable solution.

Another option would be to use an API endpoint to store/retrieve the selected repositories with a generated ID, use that ID to pass along to the return redirect, and fetch the repositories using that ID. Seems like a bit of a round trip though

**Comment by a0m0rajab** [comment]
@brandonroberts, I tried to use localstorage on middleware, but middleware is rendered on the server side, which did not solve the issue. 
Right now, the main issue is that the middleware is redirecting to the home page since the user is being logged in. Disabling it works,  but that's not a solution.

**Comment by takanome-dev** [comment]
Hey @a0m0rajab, I think I got something working...
Revert your last commit and replace the `useEffect` in `PrivateWrapper` with the following:
```tsx
useEffect(() => {
    let timeout = setTimeout(() => {
      if (isPrivateRoute && !user) {
        router.replace("/javascript/dashboard/filter/recent");
      }
    }, 1000);

    return () => {
      clearTimeout(timeout);
    };
  }, [user]);
```

If it does not work (it should work :sweat_smile:) and you're facing some rerendering, try:
```tsx
const ref = useRef<typeof user | null>(null);
ref.current = user;

useEffect(() => {
    let timeout = setTimeout(() => {
      if (isPrivateRoute && !ref.current) {
        router.replace("/javascript/dashboard/filter/recent");
      }
    }, 1000);

    return () => {
      clearTimeout(timeout);
    };
  }, [ref.current]);
```

Fingers crossed, hopefully it's gonna work :sweat_smile:

**Comment by a0m0rajab** [comment]
@takanome-dev Thank you for that, I did not try it to be honest, since I was working on adding params to the URL. I think it resolved the issue, The params I added was : `login=true`, if that's is in the URL, then the `privateWrapper` and `middleware` will not redirect to the main page. 


https://github.com/open-sauced/insights/assets/18273833/0909aa87-7f8d-46d7-bbf4-79320c8201ac

(hopefully, it works, I am not sure if something was cached while doing this lol)

File: layouts/private-wrapper.tsx [hunk 2]
```diff
@@ -10,12 +11,16 @@ interface PrivateWrapperProps {
 const PrivateWrapper = ({ isPrivateRoute = false, children }: PrivateWrapperProps) => {
   const user = useUser();
   const router = useRouter();
+  const isLoading = useStore(state => state.isLoading);
```
Comment (takanome-dev) [inline layouts/private-wrapper.tsx:14]: There is no `isLoading` in the store
```suggestion

```

File: layouts/private-wrapper.tsx [hunk 3]
```diff
@@ -10,12 +11,16 @@ interface PrivateWrapperProps {
...
   useEffect(() => {
-    if (isPrivateRoute) {
-      if (!user) router.replace("/javascript/dashboard/filter/recent");
+    if (router.asPath?.includes("login")) return;
+
+    if (isPrivateRoute && !user) {
+      router.replace("/javascript/dashboard/filter/recent");
     }
-  }, [user]);
+  }, [user, isLoading]);
```
Comment (takanome-dev) [inline layouts/private-wrapper.tsx:23]:
```suggestion
  }, [user]);
```

**Review by takanome-dev — COMMENTED** [review]
It works :+1: but I'm not sure if it's okay having the `access_token` in the URL, thoughts @brandonroberts? 

Other than that, LGTM :pizza:

File: components/organisms/Repositories/repositories.tsx [hunk 4]
```diff
@@ -55,7 +55,13 @@ const Repositories = ({ repositories }: RepositoriesProps): JSX.Element => {
         "/hub/insights/new"
       );
     } else {
-      signIn({ provider: "github" });
+      const reposIds = selectedRepos.map((repo) => repo.id);
```
Comment (OgDev-01) [inline components/organisms/Repositories/repositories.tsx:58]: nit: pass into the url a comma separated string 
```suggestion
      const reposIds = selectedRepos.map((repo) => repo.id).join(",")
```

File: components/organisms/InsightPage/InsightPage.tsx [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import { useEffect, useState } from "react";
+import { useEffect, useRef, useState } from "react";
```
Comment (OgDev-01) [inline components/organisms/InsightPage/InsightPage.tsx:1]: nit: remove unused imports
```suggestion
import { useEffect, useState } from "react";
```

File: components/organisms/Repositories/repositories.tsx [hunk 6]
```diff
@@ -55,7 +55,13 @@ const Repositories = ({ repositories }: RepositoriesProps): JSX.Element => {
...
+      signIn({ 
+        provider: "github",
+        options: {
+          redirectTo: `${window.location.origin}/hub/insights/new?selectedReposIDs=${JSON.stringify(reposIds)}&login=true&`,
```
Comment (brandonroberts) [inline components/organisms/Repositories/repositories.tsx:62]: I don't think the `login` true is needed here. We can use the `access_token` to determine if its a returned auth request.

File: lib/store/index.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ const initialState: GlobalStateInterface = {
   sessionToken: null,
   providerToken: null,
   userId: null,
-  hasReports: false
+  hasReports: false,
```
Comment (brandonroberts) [inline lib/store/index.ts:13]: Revert the changes to this file if none were made other than whitespace

File: middleware.ts [hunk 8]
```diff
@@ -12,11 +12,11 @@ export async function middleware(req: NextRequest) {
   } = await supabase.auth.getSession();
 
   // Check auth condition
-  if (session?.user) {
+  if (session?.user || req.nextUrl.searchParams.has("login")) {
```
Comment (brandonroberts) [inline middleware.ts:15]:
```suggestion
  if (session?.user || req.nextUrl.searchParams.has("access_token")) {
```

File: components/organisms/InsightPage/InsightPage.tsx [hunk 9]
```diff
@@ -41,10 +42,29 @@ const InsightPage = ({ edit, insight, pageRepos }: InsightPageProps) => {
...
+      setRepos(JSON.parse(router.query.selectedRepos as string) || []);
+    } else if (searchParams.has("selectedReposIDs")) {
+      setReposIds(JSON.parse(searchParams.get("selectedReposIDs") as string) || []);
+      searchParams.delete("login");
+      router.replace(`${pageHref.substring(0, pageHref.indexOf("?"))}?${searchParams.toString()}`);
+      setRepos(repoListData);
+    }
+  }, [repoListData, router.query.selectedRepos, pageHref]);
+
+
```
Comment (brandonroberts) [inline components/organisms/InsightPage/InsightPage.tsx:63]: Remove the extra line breaks

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: layouts/private-wrapper.tsx [hunk 10]
```diff
@@ -2,6 +2,7 @@ import { useUser } from "@supabase/auth-helpers-react";
 import { authSession } from "lib/hooks/authSession";
 import { useRouter } from "next/router";
 import React, { useEffect } from "react";
+import useStore from "lib/store";
```
Comment (NsdHSO) [inline layouts/private-wrapper.tsx:5]: Why do you need the `store` here, I think we don't need it here right now. Could you remove it ??

File: lib/hooks/useSupabaseAuth.ts [hunk 11]
```diff
@@ -39,13 +39,14 @@ const useSupabaseAuth = (loadSession = false) => {
   }, []);
 
   return {
-    signIn: (data: SignInWithOAuthCredentials) =>
-      supabase.auth.signInWithOAuth({
+    signIn: async (data: SignInWithOAuthCredentials) => {
+      await supabase.auth.signInWithOAuth({
```
Comment (NsdHSO) [inline lib/hooks/useSupabaseAuth.ts:43]: Where is the return statement??? I think we need the return statement.

**Comment by brandonroberts** [comment]
@a0m0rajab everything still works correctly when logged in. If I'm logged out, select some repos, and login, I'm redirected back to the dashboard. Looks like the middleware is catching it. Will you take a look?

**Comment by a0m0rajab** [comment]
@brandonroberts is this what you are talking about? 
I think it's an expected behavior: 

https://github.com/open-sauced/insights/assets/18273833/f29beeec-d5d4-44be-8d75-c298b4b09ba8


Or do you mean that if they logged out in the add new insight page, we need to stay at the same page?

**Comment by brandonroberts** [comment]
@a0m0rajab I was logged out, I selected some repos, and clicked "Add to Insight Page". It took me to login, and brought me back but didn't keep me on the "Create Insight Page" with the selected repos

**Comment by a0m0rajab** [comment]
@brandonroberts could it be a cache issue? I tried this with incognito mode and it worked.

https://github.com/open-sauced/insights/assets/18273833/ccda18b4-b6a8-4803-993e-2a54b617d160

**Comment by brandonroberts** [comment]
@a0m0rajab I haven't tried it locally, but only through the PR deploy preview. I tried it with incognito mode also

**Comment by a0m0rajab** [comment]
I will add a few debuggers to test the issue.

**Comment by a0m0rajab** [comment]
@brandonroberts i tested it on the preview and it worked, I had to change the remove login to signout.

**Comment by a0m0rajab** [comment]
Latest video:

https://github.com/open-sauced/insights/assets/18273833/94f7cf5e-e635-452c-a016-017db31f1656

**Review by brandonroberts — APPROVED** [review]
Minor tweak

**Review by OgDev-01 — APPROVED** [review]
